test/e2e_node/container_restart_test.go

```diff
@@ -56,8 +56,9 @@ var _ = SIGDescribe("Container Restart", feature.CriProxy, framework.WithSerial(
 		})
 
 		ginkgo.AfterEach(func() {
-			err := resetCRIProxyInjector(e2eCriProxy)
-			framework.ExpectNoError(err)
+			if err := resetCRIProxyInjector(e2eCriProxy); err != nil {
+				ginkgo.Skip("Skip the test since the CRI Proxy is undefined.")
```

Comment:
I don't understand this. Why do you skip *after* the test?

What this basically says is "something unexpected happened, let's assume that the test shouldn't have run". This looks very suspicious.

In general, skipping a test due to a runtime check is dangerous. If for whatever reason the Prow job which should run this test no longer satisfies the criteria for running the test, it'll get skipped and I bet that no-one will notice.

What you should do instead is define a `Feature:CriProxy`, label these tests with that feature, and then use label selectors in your node Prow jobs so that the test runs exactly in those jobs where it is expected to succeed.

At least that's the state for E2E testing. SIG Node is using additional tags, but I don't fully understand those.

Reply:
> I don't understand this. Why do you skip _after_ the test?

Reasoning was that skip may still call AfterEach . WDYT?

Reply:
@pohly we ran into this because we find that all of the serial node-e2e tests will pick up Feature:CriPRoxy test and we only want that test running for a the dedicated job.

I did open up https://github.com/kubernetes/test-infra/pull/33813 to explictly filter out CRIProxy from every test that this got pulled into. But it was a painful to get right and we thought that a one line change to skip was a better approach.

cc @dims

Reply:
Adding the logic to the tests is not the right direction, you already have a lot of flakiness on the e2e_node tests, you should follow the patterns in e2e tests that are way more stable ... I worked in openshift and I know this is common ... and also that CI is way more flaky than kubernetes, this is one of the reasons.
A test must be static, hermetic and idempotent, and not depend on external factors ... also avoid dynamic skips based on external variables

Reply:
> we find that all of the serial node-e2e tests will pick up Feature:CriPRoxy

Then something in your job definitions isn't right. A job should only pick tests with features that it supports. Someone needs to pick up https://github.com/kubernetes/enhancements/tree/master/keps/sig-testing/3041-node-conformance-and-features and finally figure out how E2E node testing is meant to use labels.

The problem with https://github.com/kubernetes/test-infra/pull/33813 is that you are still using regular expressions to select tests. You just cannot filter reliably with those. You have to start using `--label-filter`.

Reply:
> Reasoning was that skip may still call AfterEach , as discussed [here](https://kubernetes.slack.com/archives/C0BP8PW9G/p1732117011385279). WDYT?

Then don't use `AfterEach`?

Instead, call `ginkgo.DeferCleanup` to register something which must run if and only if the test itself ran.

Reply:
@pohly can you please share code snippet with proposal? It would be very helpfull for me, thanks.

Reply:
Putting it into `gingko.It` is what I had in mind. Adding it to `AfterEach` should indeed work, but that then would still cause it to run in skipped tests.

Reply:
Thanks @pohly ans @zhifei92 , done PTAL

Reply:
After looking at the entire file, I see that `resetCRIProxyInjector` is called also in a `BeforeEach`, combined with `Skip`.

Those `BeforeEach` calls are a bit useless right now because there's always only one `It` after them. The test setup might as well just be in that `It` call or even better, in `doTest`. I think I prefer that last approach because it avoids repetition.

If you don't like that, then at least put the `DeferCleanup` next to the code which makes the change that needs to be cleaned up, i.e. into those `BeforeEach` calls.

Reply:
Thanks @pohly , done PTAL